cpp/src/arrow/compute/exec/exec_plan.cc

```diff
@@ -170,48 +165,409 @@ Status ExecPlan::Validate() { return ToDerived(this)->Validate(); }
+                          std::vector<Expression> exprs) {
+  return input->plan()->EmplaceNode<ProjectNode>(input, std::move(label),
+                                                 std::move(exprs));
+}
+
+struct SinkNode : ExecNode {
+  SinkNode(ExecNode* input, std::string label,
+           AsyncGenerator<util::optional<ExecBatch>>* generator)
+      : ExecNode(input->plan(), std::move(label), {input}, {"collected"}, {},
+                 /*num_outputs=*/0),
+        producer_(MakeProducer(generator)) {}
+
+  static PushGenerator<util::optional<ExecBatch>>::Producer MakeProducer(
+      AsyncGenerator<util::optional<ExecBatch>>* out_gen) {
+    PushGenerator<util::optional<ExecBatch>> gen;
+    auto out = gen.producer();
+    *out_gen = std::move(gen);
+    return out;
+  }
+
+  const char* kind_name() override { return "SinkNode"; }
+
+  Status StartProducing() override { return Status::OK(); }
+
+  // sink nodes have no outputs from which to feel backpressure
+  static void NoOutputs() { DCHECK(false) << "no outputs; this should never be called"; }
+  void ResumeProducing(ExecNode* output) override { NoOutputs(); }
+  void PauseProducing(ExecNode* output) override { NoOutputs(); }
+  void StopProducing(ExecNode* output) override { NoOutputs(); }
+
+  void StopProducing() override {
+    std::unique_lock<std::mutex> lock(mutex_);
+    StopProducingUnlocked();
+  }
+
+  void InputReceived(ExecNode* input, int seq_num, ExecBatch exec_batch) override {
+    std::unique_lock<std::mutex> lock(mutex_);
+    if (stopped_) return;
+
+    // TODO would be nice to factor this out in a ReorderQueue
+    if (seq_num <= static_cast<int>(received_batches_.size())) {
+      received_batches_.resize(seq_num + 1);
+      emitted_.resize(seq_num + 1, false);
+    }
+    received_batches_[seq_num] = std::move(exec_batch);
+    ++num_received_;
+
+    if (seq_num != num_emitted_) {
+      // Cannot emit yet as there is a hole at `num_emitted_`
+      DCHECK_GT(seq_num, num_emitted_);
+      return;
+    }
+
+    if (num_received_ == emit_stop_) {
+      StopProducingUnlocked();
+    }
+
+    // Emit batches in order as far as possible
+    // First collect these batches, then unlock before producing.
+    const auto seq_start = seq_num;
+    while (seq_num < static_cast<int>(emitted_.size()) && !emitted_[seq_num]) {
+      emitted_[seq_num] = true;
+      ++seq_num;
+    }
+    DCHECK_GT(seq_num, seq_start);
+    // By moving the values now, we make sure another thread won't emit the same values
+    // below
+    std::vector<ExecBatch> to_emit(
+        std::make_move_iterator(received_batches_.begin() + seq_start),
+        std::make_move_iterator(received_batches_.begin() + seq_num));
+
+    lock.unlock();
+    for (auto&& batch : to_emit) {
+      producer_.Push(std::move(batch));
+    }
+    lock.lock();
+
+    DCHECK_EQ(seq_start, num_emitted_);  // num_emitted_ wasn't bumped in the meantime
+    num_emitted_ = seq_num;
```

Comment:
I'll rewrite this using SequencingGenerator

Reply:
@westonpace If another thread comes here, it will have a different `seq_num` and will hit the `seq_num != num_emitted_` condition above.

Reply:
I suspect the problem with putting this statement inside the lock is that (as an example) a future thread could come in while the first thread is emitting items 10 through 20 and then decide to emit item 21 before the previous thread finished.  The sequencing generator is a good workaround as it would apply the sequencing at pull time and not at push time and the generator rules prevent two threads being in that logic at once.

Reply:
In your example, `num_emitted_` is 10 until the first thread all items through 20. The second thread would simply see that 21 is larger than `num_emitted_` and return early.

Reply:
Ah, I see, you are saying that @lidavidm 's original concern is invalid.  I misread your comment, my mistake.  I agree that it wouldn't be an error as David described.  However, I still think there could be a problem at the end with the current code.  A thread could arrive to deliver the final item (let's say seq_num=100), see that `seq_num != num_emitted_` and return early.  Then the first thread finishes delivering the item with seq_num=99 and sets `num_emitted_` to 100.  At this point `StopProducing` will have already been called, no one will call `InputReceived` anymore, and the last item will be lost.

Reply:
Also, even when it isn't the final item, it could lead to unnecessary delays.  If item 10 arrives while item 9 is being delivered then it might get stuck in the buffer until item 11 arrives to unclog things.

I suppose the entire thing could be wrapped in a `while` loop that only breaks out if the next item to deliver isn't present (and the lock is held).

Reply:
You're definitely right that there should be some kind of test at the end to see if more batches need to be more emitted.